apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,726 @@
     );
 
-    promise.
-      then(() => this.emit('navigated', { panel: panel, args: args })).
-      then(nextQueuedPanel, nextQueuedPanel);
+    attachHistoryListener();
+  },
 
-    return (this.transitionPromise = promise);
+  /* will be used by tests */
+  cleanup() {
+    detachHistoryListener();
   },
 
   /**
-   * Navigates to default panel using specified arguments.
-   * @param {*?} args Any arguments that will be passed into default panel
-   * lifecycle methods.
-   * @returns {Promise} Promise that will be resolved once navigation to default
-   * panel is completed.
+   * Checks if current location.hash corresponds to default panel. This is
+   * sometimes called before we resolve the state, in that case resolve
+   * directly using findViewFromHash. Otherwise we use the resolved state.
+   *
+   * @returns {boolean} True if current location.hash corresponds to default
+   * panel.
    */
-  toDefaultPanel: function n_toDefaultPanel(args) {
-    return this.toPanel(this.defaultPanel, args);
+  isDefaultPanel() {
+    var state = currentState;
+    return state ? !state.panel.partOf : !findViewFromHash(location.hash);
   },
 
-  getPanelName: function n_getPanelName() {
-    var hash = window.location.hash;
-    if (hash.length && hash.startsWith('#')) {
-      hash = hash.slice(1);
-    }
-
-    var index;
-    if ((index = hash.indexOf('&')) !== -1) {
-      hash = hash.slice(0, index);
-    }
-
-    return hash;
+  /**
+    * Called with a first argument that is not an object, it returns whether the
+    * current panel is this first argument.
+    *
+    * Called with 2 arguments, the first one is not an object, the second one
+    * must be an object. It returns whether the current panel is the first
+    * argument, and for each property of the second argument, whether its value
+    * is equal to the value for the current panel's arguments' same property.
+    *
+    * @param {String} panelName Checks if the current panel is this panelName.
+    * @param {Object} [args] Checks against the current panel enter
+    * arguments.
+    * @returns {boolean} Result of the comparison.
+    */
+  isCurrentPanel(panelName, args) {
+    // TODO compare args
+    return currentState && currentState.panel.name === panelName;
   },
 
-  slide: function n_slide(position) {
-    var wrapper = this.mainWrapper;
-
-    if (wrapper.dataset.position === position) {
-      return Promise.resolve();
-    }
-
-    wrapper.dataset.position = position;
-
-    return new Promise(function(resolve, reject) {
-      // We have 2 panels, so we get 2 transitionend for each step
-      var trEndCount = 0;
-
-      wrapper.addEventListener('transitionend', function trWait(e) {
-        trEndCount++;
-
-        if (trEndCount < 2) {
-          return;
-        }
+  once(event, handler) {
+    // TODO
+    setTimeout(handler);
+  },
 
-        e.currentTarget.removeEventListener(e.type, trWait);
-        resolve();
-      });
-    });
-  }
+  setReady() {}
```

Comment:
Ideally I'd like to remove these 2 methods, I'm not sure they're still useful -- but I might be wrong.